projects/steering/test/test_steering.c

```diff
@@ -0,0 +1,109 @@
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include "adc.h"
+#include "adc_periodic_reader.h"
+#include "can.h"
+#include "can_msg_defs.h"
+#include "delay.h"
+#include "event_queue.h"
+#include "exported_enums.h"
+#include "gpio.h"
+#include "gpio_it.h"
+#include "interrupt_def.h"
+#include "misc.h"
+#include "ms_test_helpers.h"
+#include "soft_timer.h"
+#include "status.h"
+#include "steering_can.h"
+#include "steering_control_stalk.h"
+#include "steering_digital_input.h"
+#include "steering_events.h"
+#include "test_helpers.h"
+#include "wait.h"
+
+#define STEERING_CAN_DEVICE_ID 0x1
+
+typedef enum {
+  STEERING_CAN_EVENT_RX = 10,
+  STEERING_CAN_EVENT_TX,
+  STEERING_CAN_FAULT,
+} SteeringCanEvent;
+
+CanSettings can_settings = {
+  .device_id = STEERING_CAN_DEVICE_ID,
+  .bitrate = CAN_HW_BITRATE_125KBPS,
+  .rx_event = STEERING_CAN_EVENT_RX,
+  .tx_event = STEERING_CAN_EVENT_TX,
+  .fault_event = STEERING_CAN_FAULT,
+  .tx = { GPIO_PORT_A, 12 },
+  .rx = { GPIO_PORT_A, 11 },
+};
+
+static CanStorage s_can_storage;
+
+void setup_test(void) {
+  adc_init(ADC_MODE_SINGLE);
+  gpio_init();
+  interrupt_init();
+  event_queue_init();
+  gpio_it_init();
+  soft_timer_init();
+  steering_digital_input_init();
+  can_init(&s_can_storage, &can_settings);
+  adc_periodic_reader_init();
+  control_stalk_init();
+}
+
+void test_steering_digital_input_horn() {
+  GpioAddress *horn_address = test_get_address(STEERING_INPUT_HORN_EVENT);
+  TEST_ASSERT_OK(gpio_it_trigger_interrupt(horn_address));
+
+  Event e = { 0 };
+  TEST_ASSERT_OK(event_process(&e));
+  TEST_ASSERT_EQUAL(STEERING_INPUT_HORN_EVENT, e.id);
+  TEST_ASSERT_EQUAL(GPIO_STATE_LOW, e.data);
+  // Should be empty after the event is popped off

```

Comment:
use this [cool-ass](https://github.com/uw-midsun/firmware_xiv/blob/5d77d33663df6ce34f54f20b006a234af94a889a/libraries/ms-common/inc/ms_test_helpers.h#L71) macro: MS_TEST_HELPER_ASSERT_NO_EVENT_RAISED